packages/asset-swapper/src/utils/market_operation_utils/balancer_utils.ts

```diff
@@ -116,36 +118,75 @@ export class BalancerPoolsCache {
             return [];
         }
     }
-}
 
-// tslint:disable completed-docs
-export function computeBalancerSellQuote(pool: BalancerPool, takerFillAmount: BigNumber): BigNumber {
-    if (takerFillAmount.isGreaterThan(bmath.bmul(pool.balanceIn, bmath.MAX_IN_RATIO))) {
-        return ZERO_AMOUNT;
-    }
-    const weightRatio = pool.weightIn.dividedBy(pool.weightOut);
-    const adjustedIn = bmath.BONE.minus(pool.swapFee)
-        .dividedBy(bmath.BONE)
-        .times(takerFillAmount);
-    const y = pool.balanceIn.dividedBy(pool.balanceIn.plus(adjustedIn));
-    const foo = Math.pow(y.toNumber(), weightRatio.toNumber());
-    const bar = new BigNumber(1).minus(foo);
-    const tokenAmountOut = pool.balanceOut.times(bar);
-    return tokenAmountOut.integerValue();
-}
+    protected async _loadTopPoolsAsync(): Promise<void> {
+        const fromToPools: {
+            [from: string]: { [to: string]: Pool[] };
+        } = {};
 
-export function computeBalancerBuyQuote(pool: BalancerPool, makerFillAmount: BigNumber): BigNumber {
-    if (makerFillAmount.isGreaterThan(bmath.bmul(pool.balanceOut, bmath.MAX_OUT_RATIO))) {
-        return ZERO_AMOUNT;
+        const pools = await this._fetchTopPoolsAsync();
+        pools.forEach(pool => {
+            const { tokensList } = pool;
+            for (const from of tokensList) {
+                for (const to of tokensList.filter(t => t.toLowerCase() !== from.toLowerCase())) {
+                    if (!fromToPools[from]) {
+                        fromToPools[from] = {};
+                    }
+                    if (!fromToPools[from][to]) {
+                        fromToPools[from][to] = [];
+                    }
+                    try {
+                        // The list of pools must be relevant to `from` and `to`  for `parsePoolData`
+                        const poolData = parsePoolData([pool], from, to);
+                        fromToPools[from][to].push(poolData[0]);
+                        // Cache this as we progress through
+                        this._cachePoolsForPair(from, to, fromToPools[from][to]);
+                    } catch {
+                        // soldier on
+                    }
+                }
+            }
+        });
     }
-    const weightRatio = pool.weightOut.dividedBy(pool.weightIn);
-    const diff = pool.balanceOut.minus(makerFillAmount);
-    const y = pool.balanceOut.dividedBy(diff);
-    let foo: number | Decimal = Math.pow(y.toNumber(), weightRatio.toNumber()) - 1;
-    if (!Number.isFinite(foo)) {
-        foo = new Decimal20(y.toString()).pow(weightRatio.toString()).minus(1);
+
+    protected async _fetchTopPoolsAsync(): Promise<BalancerPoolResponse[]> {
```

Comment:
Like somewhere around here?

Reply:
on a cache miss it reloads just `taker->Maker` and keeps `maxPoolsFetched` of the pools in memory. Currently set to 3